### PR TITLE
Fix for mypy 1.4 changes

### LINKED ISSUE #84: Breaking change in `TypeVarType` causes an error during typecheck
Mypy 1.4 includes PR https://github.com/python/mypy/pull/14872 that added a positional `deault` parameter to the `mypy.types.TypeVarType` constructor. `plugin.py::takes_callable_and_args_callback()` invokes it without that argument, causing `mypy` to crash. 

This seems to be fixed by passing `AnyType(TypeOfAny.from_omitted_generics)` as an additional argument, but this will make `trio-typing` incompatible with `mypy<1.4`. Thoughts on how to proceed?

### LINKED ISSUE #85: Incompatible with `trio >= 0.22.1` due to `async_generator`
see https://github.com/python-trio/trio/commit/a58a58c57f2c7015851c0510a7822c3b797ab71a

`async_generator` is no longer a dependency of `trio`, hence https://github.com/python-trio/trio-typing/blob/8c3677873c7fb43781bb40820c44d0ba1f3ebe25/trio_typing/__init__.py#L5 causes an `ImportError`

### PR DESCRIPTION
**Pull request by VincentVanlaer**
This PR contains all the necessary changes to the plugin, tests and stubs for mypy 1.4

## Api changes

- See https://github.com/python/mypy/pull/14872 and https://github.com/python/mypy/pull/13546 for API changes
- Tests have been updated as the output format of mypy has changed a bit

## Stubtest changes

As stubtest is part of mypy, and therefore also got updated, the stubs required signficant changes:

#### List of relevant changes:
- A whole bunch of types that are final at runtime, are now final in
  for type checking purposes as well
- Add missing attrs decorators to some types. Internal attrs members
  have been added to the allowlist
- StrictFIFOLock is no longer a subclass of Lock, as Lock is suposed to
  be final.
- Converted attributes to read-only properties whenever this was
  necessary to match trio

#### Minor changes:
- Ignore internal APIs and tests (previously seemed to be the case by
  default)
- Typevars and aliases have been prefixed with underscores, so that they
  are not part of the public interface of the modules (as they are not
  available at runtime)

Fixes #84 
Fixes #85

### CODE REVIEW

**Comment by A5rocks** [comment]
Could you check the mypy version we are passed to choose whether to pass `default=` or not?

**Comment by VincentVanlaer** [comment]
I have added a check for the mypy version (and some other required changes due to changes in trio)

**Review by A5rocks — APPROVED** [review]
Assuming the type changes are correct, this looks fine! I'd like someone who actually knows `trio-typing` to look at this before merging though :P

File: setup.py
Comment (Redoubts) [inline setup.py:26]: Why do you need this backport package?

Comment (CoolCat467) [reply setup.py:26]: Because otherwise you have this error: `Error importing plugin "trio_typing.plugin": No module named 'async_generator'`

Comment (VincentVanlaer) [reply setup.py:26]: See #85

**Review by jakkdl — APPROVED** [review]
I'm no expert on trio-typing (is anybody?), but seems fine to me.

**Review by TeamSpen210 — APPROVED** [review]

File: trio-stubs/lowlevel.pyi
Comment (oremanj) [inline trio-stubs/lowlevel.pyi:182]: I don't think stubtest should be insisting that we expose underscore-prefixed attributes...

**Review by oremanj — APPROVED** [review]
Sorry for the delay in looking at this. I'm confused about one of the changes but that doesn't need to block merging.

File: trio-stubs/lowlevel.pyi
Comment (CoolCat467) [reply trio-stubs/lowlevel.pyi:182]: The idea is that these attributes don't exist in the real thing, but if we say they do in the stubs, typecheckers think that they do exist, which confuses everyone. It is best practice to underscore thinks that don't exist in the real version.

Comment (VincentVanlaer) [reply trio-stubs/lowlevel.pyi:182]: It's not really stubtest's fault here. It needs these attributes to construct the correct `__init__` with attrs. The error only complains about `__init__` and `__match_args__`, not missing attributes. I can remove the `_NO_DEFAULT` and just inline the `object()` however.

**Comment by VincentVanlaer** [comment]
Thanks for all the reviews!

**Pull request merged**
